### PR TITLE
Used MVVM approach for ConnectProviderFragment

### PR DESCRIPTION
**Pull request by mnewlive**
**Related Issue**  
https://github.com/saltedge/sca-authenticator-android/issues/134

**Screenshots (if applicable)**  
image
image
image
image
image
image
image
image

### CODE REVIEW

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderFragment.kt [hunk 1]
```diff
@@ -42,19 +47,21 @@ import kotlinx.android.synthetic.main.fragment_connect.*
...
     private val webViewClient = ConnectWebClient(contract = this)
+    private lateinit var binding: ConnectProviderBinding
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        injectDependencies()
-        presenterContract.setInitialData(
+        authenticatorApp?.appComponent?.inject(this)
+        setupViewModel()
+        viewModel.setInitialData(
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderFragment.kt:64]: move to `setupViewModel`

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderFragment.kt [hunk 2]
```diff
@@ -66,73 +73,54 @@ class ConnectProviderFragment : BaseFragment(),
...
+        viewModel.onViewClick(view?.id ?: return)
+    }
+
+    override fun onClick(listener: DialogInterface?, dialogActionId: Int) {
+        viewModel.onDialogActionIdClick(dialogActionId)
     }
 
     override fun onBackPress(): Boolean {
-        return if (presenterContract.shouldShowWebView && connectWebView?.canGoBack() == true) {
+        return if (viewModel.shouldShowWebView() && connectWebView?.canGoBack() == true) {
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderFragment.kt:106]: move business logic to ViewModel

File: app/src/main/kotlin/com/saltedge/authenticator/features/onboarding/OnboardingSetupViewModel.kt [hunk 3]
```diff
@@ -39,21 +39,12 @@ class OnboardingSetupViewModel(
...
-
-    init {
-        headerTitle.value = R.string.onboarding_secure_app_passcode_create
-        setupLayoutVisibility.value = View.GONE
-        onboardingLayoutVisibility.value = View.VISIBLE
-        proceedViewVisibility.value = View.GONE
-        skipViewVisibility.value = View.VISIBLE
-        passcodeInputViewVisibility.value = View.VISIBLE
-    }
+    val headerTitle: MutableLiveData<Int> = MutableLiveData(R.string.onboarding_secure_app_passcode_create )
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/onboarding/OnboardingSetupViewModel.kt:42]: redundant space after `R.string.onboarding_secure_app_passcode_create`

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt [hunk 4]
```diff
@@ -43,101 +49,34 @@ import com.saltedge.authenticator.sdk.tools.keystore.KeyStoreManagerAbs
...
     private val connectionsRepository: ConnectionsRepositoryAbs,
     private val keyStoreManager: KeyStoreManagerAbs,
     private val apiManager: AuthenticatorApiManagerAbs
-) : ConnectProviderContract.Presenter, ConnectionCreateListener, FetchProviderConfigurationListener {
+) : ViewModel(), LifecycleObserver, ConnectionCreateListener, FetchProviderConfigurationListener {
 
-    override val reportProblemActionText: Int?
-        get() = if (viewMode.isCompleteWithSuccess) null else R.string.actions_contact_support
-    private var sessionFailMessage: String? = null
+    val iconResId: MutableLiveData<Int> = MutableLiveData(R.drawable.ic_status_error)
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt:60]: variable name is out of context.

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt [hunk 5]
```diff
@@ -43,101 +49,34 @@ import com.saltedge.authenticator.sdk.tools.keystore.KeyStoreManagerAbs
...
-            else -> viewContract?.showErrorAndFinish(appContext.getString(R.string.errors_unable_connect_provider))
-        }
-    }
-
-    override fun onConnectionCreateFailure(error: ApiErrorData) {
-        viewContract?.showErrorAndFinish(error.getErrorMessage(appContext))
-    }
+    var onCloseEvent = MutableLiveData<ViewModelEvent<Unit>>()
+        private set
+    var showErrorAndFinishEvent = MutableLiveData<ViewModelEvent<String>>()
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt:76]: this variable is intended only to show error. rename it

Comment (ConstantinKV) [reply app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt:76]: it is event. don't use imperative

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt [hunk 6]
```diff
@@ -43,101 +49,34 @@ import com.saltedge.authenticator.sdk.tools.keystore.KeyStoreManagerAbs
...
+        private set
+    var loadUrlInWebViewEvent = MutableLiveData<ViewModelEvent<String>>()
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt:78]: it is event. don't use imperative. don't indicate where observer should load URL

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt [hunk 7]
```diff
@@ -213,16 +208,78 @@ class ConnectProviderPresenter @Inject constructor(
...
+                R.string.errors_connection_failed_description
+            ) ?: ""
+        }
+    }
+
+    private fun getActionTextResId(): Int {
+        return if (viewMode.isCompleteWithSuccess) R.string.actions_proceed else R.string.actions_try_again
+    }
+
+    private fun showWebView(): Int {
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt:272]: this functions show nothing. rename it

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt [hunk 8]
```diff
@@ -158,49 +97,105 @@ class ConnectProviderPresenter @Inject constructor(
...
+
+    fun onViewClick(viewId: Int) {
+        if (viewId == R.id.mainActionView) onCloseEvent.postValue(ViewModelEvent(Unit))
+    }
+
+    fun onDialogActionIdClick(dialogActionId: Int) {
+        if (dialogActionId == DialogInterface.BUTTON_POSITIVE) onCloseEvent.postValue(ViewModelEvent(Unit))
+    }
+
+    fun webAuthFinishSuccess(id: ConnectionID, accessToken: Token) {
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderViewModel.kt:180]: redundant function. call directly authFinishedWithSuccess

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderFragment.kt [hunk 9]
```diff
@@ -66,73 +73,54 @@ class ConnectProviderFragment : BaseFragment(),
...
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
         connectWebView?.webViewClient = webViewClient
         completeView?.setOnClickListener(this)
-        updateViewsContent()
-        presenterContract.viewContract = this
-        presenterContract.onViewCreated()
+        viewModel.onViewCreated()
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderFragment.kt:89]: maybe better to move to `onResume`?

File: app/src/main/res/layout/view_empty.xml [hunk 10]
```diff
@@ -35,31 +35,32 @@
...
         android:layout_marginEnd="@dimen/dp_32"
-        android:layout_marginTop="@dimen/dp_12"
         android:gravity="center_horizontal"
         android:maxLines="5"
         tools:text="Description text view" />
 
     <Button
         android:id="@+id/actionView"
         style="@style/LargeBlueButton"
+        android:layout_height="@dimen/dp_54"
```
Comment (ConstantinKV) [inline app/src/main/res/layout/view_empty.xml:63]: as i see in design all blue buttons have height = 48dp

File: app/src/main/res/layout/fragment_connect.xml [hunk 11]
```diff
@@ -18,22 +18,37 @@
...
-</FrameLayout>
+        android:layout_height="match_parent">
+
+        <com.saltedge.authenticator.sdk.web.SEWebView
+            android:id="@+id/connectWebView"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:visibility="@{viewModel.shouldShowWebViewVisibility}" />
+
+        <include
```
Comment (ConstantinKV) [inline app/src/main/res/layout/fragment_connect.xml:42]: fragment_connect_processing is equal to complete_view. why we need it?

File: app/src/main/kotlin/com/saltedge/authenticator/features/onboarding/OnboardingSetupViewModel.kt [hunk 12]
```diff
@@ -39,21 +39,12 @@ class OnboardingSetupViewModel(
...
+    val proceedViewVisibility: MutableLiveData<Int> = MutableLiveData(View.GONE)
+    val skipViewVisibility: MutableLiveData<Int> = MutableLiveData(View.VISIBLE)
+    val setupLayoutVisibility: MutableLiveData<Int> = MutableLiveData(View.GONE)
+    val onboardingLayoutVisibility: MutableLiveData<Int> = MutableLiveData(View.VISIBLE)
+    val passcodeInputViewVisibility: MutableLiveData<Int> = MutableLiveData(View.VISIBLE)
 
     var pageIndicator = MutableLiveData<Int>()
```
Comment (ConstantinKV) [inline app/src/main/kotlin/com/saltedge/authenticator/features/onboarding/OnboardingSetupViewModel.kt:49]: val

Comment (mnewlive) [reply app/src/main/kotlin/com/saltedge/authenticator/features/onboarding/OnboardingSetupViewModel.kt:49]: i used `private set`

File: app/src/main/res/layout/fragment_connect.xml (hunk 11 above)
Comment (mnewlive) [reply app/src/main/res/layout/fragment_connect.xml:42]: The design is still being updated, so I think we can leave it like this for now, so as not to delay this `pr`

**Pull request merged**